Unidirectional cluster links (#18195)
canonical/lxd: doc/explanation/clusters.md

(exp-cluster-links)=
## Cluster links

Cluster links enable communication between separate LXD clusters by pinning the remote cluster's TLS certificate and optionally establishing mutual trust.

Cluster links are the foundation for {ref}`replicators <exp-replicators>`, which use bidirectional links to sync instances across clusters for active-passive disaster recovery.
Unidirectional links are suited to scenarios where one cluster needs to read from another without granting reciprocal access.

### Link types

There are two link types, each suited to different trust and access requirements:

`bidirectional`
: Both clusters authenticate each other using mutual TLS. Each side creates an identity for the other and can initiate requests to the other. This is the default type.

`unidirectional`

> : A pins B's certificate and uses a token to activate a pending identity that B created for A. B stores a corresponding link entry and can authenticate incoming requests from A, but holds no address for A and cannot initiate requests to it.

Reviewer: At present, you introduce each cluster link type by describing the connection process. How about beginning by describing the *relationship* between the two clusters? This seems to be the main quality that distinguishes the different types of cluster links.

Also, this is the first time that you use A/B, so the word "Cluster" should be included. Additionally, I would recommend clarifying the difference between Cluster A and Cluster B here at the outset (Cluster A only sends, Cluster B only receives).

How about:

```suggestion
`bidirectional`
: Either cluster can initiate requests to the other cluster.
The clusters authenticate each other using mutual TLS, and both clusters create an identity for the other side.
This is the default type.

`unidirectional`
: Requests can only be sent in one direction: from Cluster A to Cluster B.
Cluster A pins B's certificate and uses a token to activate a pending identity that B created for A.
B stores a corresponding link entry and can authenticate incoming requests from A, but holds no address for A and cannot initiate requests to it.
```

### Connection process

> All link types rely on TLS certificate pinning: A fetches and pins B's certificate before making any connection.

Reviewer: ```suggestion
All link types rely on TLS certificate pinning: Cluster A fetches and pins Cluster B's certificate before making any connection.
```

The link type determines what trust is established on B's side.

#### Bidirectional connection process

1. **Cluster A** creates a pending cluster link and generates a trust token.
1. **Cluster B** uses this token to establish the connection and send its certificate back.
1. Both clusters validate certificates and activate the bidirectional link.
1. The link becomes active and both clusters can communicate.

#### Unidirectional connection process

> 1. **Cluster B** issues a pending identity token using `lxc auth identity create cluster-link/<name>`.

Reviewer 2: ```suggestion
1. **Cluster B** issues a pending identity token using [`lxc auth identity create`](lxc_auth_identity_create.md) `cluster-link/<name>`.
```

> 1. **Cluster A** consumes the token with `lxc cluster link create <name> --token <token> --unidirectional`, pins B's certificate, and calls back to B to activate the pending identity.

Reviewer 2: ```suggestion
1. **Cluster A** consumes the token with [`lxc cluster link create`](lxc_cluster_link_create.md) `<name> --token <token> --unidirectional`, pins B's certificate, and calls back to B to activate the pending identity.
```

> 1. A has an active link to B with no associated identity. B has an active identity for A and a corresponding link row.

Reviewer: ```suggestion
1. Cluster A has an active link to B with no associated identity.
Cluster B has an active identity for A and a corresponding link row.
```

For more information, see: {ref}`howto-cluster-links-create`.

(exp-clusters-links-identity)=
### Identity management

The identities created depend on the link type:

- **Bidirectional**: LXD creates a `Cluster link certificate` identity on each side. The identity can be in one of two states:
  - **Pending**: A trust token has been generated but the link has not been activated yet.
  - **Active**: Both clusters have exchanged certificates and the link is operational.

> - **Unidirectional**: B creates an identity for A and a corresponding link entry (no addresses, so B cannot reach A, but can manage A's access via `lxc cluster link delete`).

Reviewer 2: ```suggestion
- **Unidirectional**: B creates an identity for A and a corresponding link entry (no addresses, so B cannot reach A, but can manage A's access via [`lxc cluster link delete`](lxc_cluster_link_delete.md)).
```

> - **Unidirectional**: B creates an identity for A and a corresponding link entry (no addresses, so B cannot reach A, but can manage A's access via [`lxc cluster link delete`](lxc_cluster_link_delete.md)).

Reviewer: ```suggestion
- **Unidirectional**: Cluster B creates an identity for Cluster A and a corresponding link entry (no addresses, so B cannot reach A, but can manage A's access via [`lxc cluster link delete`](lxc_cluster_link_delete.md)).
```

>   A stores B's certificate directly without an associated identity.

Reviewer: ```suggestion
Cluster A stores B's certificate directly without an associated identity.
```

Identities are managed using {ref}`fine-grained authorization <fine-grained-authorization>`.

### Security considerations

- **Certificate validation**: All connections verify certificate fingerprints.
- **Fine-grained permissions**: Linked clusters can be granted specific entitlements (for example, only backup operations).
- **Identity isolation**: Each cluster link gets its own identity that can be managed independently.
- **Group membership**: Cluster link identities can be assigned to authentication groups for bulk permission management.

Together, these controls limit the potential impact of a compromised link by enforcing certificate-based trust and least-privilege access.
They also make it possible to revoke a single link's access without impacting other cluster-to-cluster trust relationships.

### Removal

Deletion behavior depends on the link type:

> - **Bidirectional**: Run `lxc cluster link delete` on both clusters to fully remove the trust relationship.

Reviewer 2: ```suggestion
- **Bidirectional**: Run [`lxc cluster link delete`](lxc_cluster_link_delete.md) on both clusters to fully remove the trust relationship.
```

- **Unidirectional**: Deleting on A removes only A's link.

>   B's identity and link row remain until B explicitly deletes its link.

Reviewer: You have this information in the how-to guide as well, with almost exactly the same wording for the "unidirectional" link. Given that, I think you might simply remove this from the explanation page. The language seems better suited to the how-to guide, and you can add a sentence to the cluster link section in the explanation page along the lines of:

```
Refer to our {ref}`guide to managing cluster links <howto-cluster-links-manage>` for more information about how to view, configure, and delete cluster links.
```

Author: Revoking trust via removal is important to include in the explanation page too. We're trading brevity for increased awareness around security implications.

Reviewer: I'd recommend using the word "security" here, then, and explicitly stating the security implications. At present, this reads more as a short guide to how to remove the links, rather than a warning about security. We might revisit later, too, and place these details into a "warning" admonition in the "Security considerations" subsection.

### Member status

A cluster link member can have one of the following statuses.

> You can see member status with `lxc cluster link info` (while `lxc cluster link list` shows the link identity status and link type); see {ref}`howto-cluster-links-view`.

Reviewer 2: ```suggestion
You can see member status with [`lxc cluster link info`](lxc_cluster_link_info.md) (while [`lxc cluster link list`](lxc_cluster_link_list.md) shows the link identity status and link type); see {ref}`howto-cluster-links-view`.
```

> You can see member status with [`lxc cluster link info`](lxc_cluster_link_info.md) (while [`lxc cluster link list`](lxc_cluster_link_list.md) shows the link identity status and link type); see {ref}`howto-cluster-links-view`.

Reviewer: To align with inclusive language standards, use words such as "refer to", "consult", "access", "visit", etc.rather than "see". Also, why is it necessary to include the identity status command here? It does not seem to relate to the section topic: member status. Instead, how about expanding on the cross reference, as the how-to guide will include that info?

```suggestion
Run [`lxc cluster link info`](lxc_cluster_link_info.md) to check member status.
(Refer to {ref}`howto-cluster-links-view` for additional details.)
```

Author: Ack on inclusive language standards, thanks for raising that!

Identities and cluster links are closely related so I'd like to include the identity status command here.

Reviewer: As I understand it, the change to "refer to", etc. from "see" is a recent addition to the style guide, that the documentation hasn't implemented yet, but I'm trying to future-proof new PRs by catching these words when they appear.

As for identities and cluster links: if they're related, this would be a good place to spell out that relationship. At present, when I read the parenthetical, I wonder, "why do I need to know this, when this is a section about 'Member status'?" I'd also move that sentence about identities to after the bullet-point list. Keep the user focused on member status until they've read through the different statuses, and then introduce new information.

Author: Sounds good, I'll reword that section and add more info on the relationship.

(exp-cluster-links)=
## Cluster links

Cluster links enable communication between separate LXD clusters by pinning the remote cluster's TLS certificate and optionally establishing mutual trust.

> Cluster links are the foundation for {ref}`replicators <exp-replicators>`, which use bidirectional links to sync instances across clusters for active-passive disaster recovery. Unidirectional links are suited to scenarios where one cluster needs to read from another without granting reciprocal access.

Reviewer 3: > Unidirectional links are suited to scenarios where one cluster needs to read from another without granting reciprocal access

Lets drop the "read" and replace with "access another cluster" without granting reciprocal access, as there's no reason it can only be used for reads.

### Link types

There are two link types, each suited to different trust and access requirements:

`bidirectional`
: Either cluster can initiate requests to the other cluster. The clusters authenticate each other using mutual TLS, and both clusters create an identity for the other side. This is the default type.

`unidirectional`

> : Requests can only be sent in one direction: from Cluster A to Cluster B. Cluster A pins B's certificate and uses a token to activate a pending identity that B created for A. B stores a corresponding link entry and can authenticate incoming requests from A, but holds no address for A and cannot initiate requests to it.

Reviewer 3: > B stores a corresponding link entry and can authenticate incoming requests from A, but holds no address for A and cannot initiate requests to it.

What is a "link entry"? My understanding is that B does not have a cluster-link record, but only a cluster-link identity, if so we should be clear here what we mean.

Author: That's correct, I've clarified this in the docs.

### Connection process

> All link types rely on TLS certificate pinning: Cluster A fetches and pins Cluster B's certificate before making any connection. The link type determines what trust is established on Cluster B's side.

Reviewer 3: > The link type determines what trust is established on Cluster B's side.

Suggest changing to:

> The link type determines the level (and direction) of trust that is established between Cluster A and Cluster B.

Identities are managed using {ref}`fine-grained authorization <fine-grained-authorization>`.

### Security considerations

- **Certificate validation**: All connections verify certificate fingerprints.
- **Fine-grained permissions**: Linked clusters can be granted specific entitlements (for example, only backup operations).
- **Identity isolation**: Each cluster link gets its own identity that can be managed independently.
- **Group membership**: Cluster link identities can be assigned to authentication groups for bulk permission management.

Together, these controls limit the potential impact of a compromised link by enforcing certificate-based trust and least-privilege access. They also make it possible to revoke a single link's access without impacting other cluster-to-cluster trust relationships.

### Removal

Deleting a cluster link revokes the security trust it established. The scope depends on the link type:

- **Bidirectional**: Run [`lxc cluster link delete`](lxc_cluster_link_delete.md) on both clusters to fully remove the trust relationship.

> - **Unidirectional**: Deleting on Cluster A removes only A's link. Cluster B's identity and link row remain until B explicitly deletes its link.

Reviewer 3: There is no link on Cluster B to delete right?

Reviewer: A quick note: this kind of issue with the terminology was something I also wanted to revisit at some point. It seems to me that there's a kind of fuzziness with the word "link."

In some parts of the docs, it seems to refer to "something" that exists on a single cluster, as in this example, where B has to "explicitly delete its link". 

In other parts of the docs, the word "link" seems to refer to several "things" (identities, link rows, certificates, etc.) that exist on one or more clusters and that *collectively* make up a link. For example: "Both clusters validate certificates and activate the bidirectional link." (The singular "bidirectional link" makes it sound as if the "link" is a something that spans the two clusters, rather than two different links, one on each cluster.) Similarly, in discussing replicators, "A cluster link must be established between the two clusters." Again, a single link between clusters, rather than two links, one on each cluster.

When talking about a cluster link and a single cluster, does the word "link" refer to something separate from a "link row" and/or "link identity"? Does it make sense to refer to cluster links in the plural when talking about two clusters (e.g. to set up a replicator, we first create bidirectional cluster links between Cluster A and Cluster B)?

Reviewer 3: A Cluster Link is a named concept in LXD. For bidirectional links there is a Cluster Link entity defined on each cluster.

For unidirectional and public links there is only a Cluster Link defined on the "A" side.

For unidirectional links the "B" side just has an Identity entity defined for the "A" side to authenticate against.

So we need to differentiate between "Cluster Link" (the entity, which may exist on one or both sides) and the conceptual "link" trust relationship we are setting up.

Reviewer: Thanks for the clarification, @tomponline . I can see where some of my confusion was coming from, now. As an example, when I read the section of the guide on how to "delete a cluster link," I thought this was about removing the "conceptual link" rather than the Cluster Link entity, and I expected the full sequence of steps to remove the trust relationship, rather than a description of the effects of the `delete` command.

One thought on differentiating the two: we might consistently use "trust relationship" (already used in a few places) rather than "link" to refer to the conceptual link.

For example, in the last stages of the bilateral connection process:

```
1. Both clusters validate certificates and activate the bidirectional link.
1. The link becomes active and both clusters can communicate.
```

This could become:

```
1. Both clusters validate certificates and activate their links.
1. The trust relationship is established and both clusters can communicate.
```

Reviewer 3: Yes that would help disambiguate things :)

Author: I've reworded the docs to better differentiate between cluster links and conceptual links.
